### ISSUE TITLE
Some UI screenshot tests fail at random 

### ISSUE THREAD

**Issue by mattab** [piwik/piwik#6693]
This issue is to keep track of some of the UI screenshot tests that randomly fail. Hopefully we can solve of some of those issues and identify best practise for future UI tests.

Note that we already had to disable a few tests because they failed randomly. They are marked with `it.skip` there are listed here: https://github.com/piwik/piwik-ui-tests/search?utf8=%E2%9C%93&q=%22it.skip%28%22&type=Code

**Comment by mattab** [comment]
* 1 x ViewDataTable few pixel differences [build 10354](http://builds-artifacts.piwik.org/ui-tests.master/10354.8/screenshot-diffs/singlediff.html?processed=../processed-ui-screenshots/ViewDataTableTest_10_change_limit.png&expected=ViewDataTableTest_10_change_limit.png&github=ViewDataTableTest_10_change_limit.png)

**Comment by mattab** [comment]
The Transitions random failure was fixed. Now we often see this error: http://builds-artifacts.piwik.org/ui-tests.master/10667.7/screenshot-diffs/singlediff.html?processed=../processed-ui-screenshots/ActionsDataTable_column_sorted.png&expected=ActionsDataTable_column_sorted.png&github=ActionsDataTable_column_sorted.png

but unfortunately we still don't know how to prevent this random failure. Must be some crazy edge case in Javascript...

**Comment by mattab** [comment]
Moving to `Short term` because now random UI test failure is almost the only reason for failed CI builds. Needed for #6544

**Comment by mattab** [comment]
moving to 2.12.0 to try do something serious about this issue - we simply **need** green builds in the morning when starting to work :-)

**Comment by mattab** [comment]
There is some feature in mocha to retry tests: https://www.npmjs.com/package/mocha-retry

we agree with what the devs says eg. `Why you shouldn't use this? Tests that pass sometimes, and sometimes they don't are a bad smell. Your tests should be consistent: either they always pass, or they always fail. So, don't you this Mocha plugin, instead make your tests consistent.`

however in this case i reckon none of us know how to fix the UI tests, so maybe a retry (only for those tests that are know to randomly fail) would be a correct solution

**Comment by mattab** [comment]
I asked the developer of mocha-retry for help in https://github.com/giggio/mocha-retry/issues/6

If we can't make mocha-retry work maybe we could try this other tool: https://github.com/lennym/mocha-extra-shot

**Comment by mattab** [comment]
Answer from the dev of mocha-retry:

> Yes, my version worked, but broke the browser, they said. I haven't updated it since then. They said they would look into it, but never did. Maybe you would like to help and see what went wrong there? I never got to it, and now I am on a project on a different technology, I will not find time to work on this for quite a while.

So it looks like we're back to square one, we don't have a way to retry the screenshot tests. maybe the only way will be to actually solve the random screenshot rendering issues...

**Comment by mattab** [comment]
I noticed that Visitor Log now has random failures in Screenshot tests as well since we changed how the `#x [icon]` (for goals) is positioned. 

[example random fail](http://builds-artifacts.piwik.org/ui-tests.master/11646.7/screenshot-diffs/singlediff.html?processed=../processed-ui-screenshots/UIIntegrationTest_ecommerce_log.png&expected=UIIntegrationTest_ecommerce_log.png&github=UIIntegrationTest_ecommerce_log.png)

it would be nice not to add a new randomly failing test case as we have troubles with those...  @sgiehl  do you maybe have an idea of how we could change some CSS to avoid such failure? it was changed iirc in this commit https://github.com/piwik/piwik/commit/aac931602cb756a10951205a8dfb045b14912433

**Comment by sgiehl** [comment]
We are using phantomjs (+phantomcss/casperjs) in anther project. There are similar issues with elements not rendering the same. In some cases it helped to wait 100ms before taking the screenshot or let phantomjs scroll to bottom and to top again. That's a bit creepy. Maybe that issues might get better with using phantomjs 2 (as soon as a linux package is available)

**Comment by mattab** [comment]
2 solutions to fix the UI tests:
 * either fix the core CoSS / JS problem if we can identify it (maybe difficult?)
 * or implement a "retry" mechanism for those tests that are randomly failing...

**Comment by mattab** [comment]
Here is a typical random screenshot failure: http://builds-artifacts.piwik.org/ui-tests.master/13257.7/screenshot-diffs/diffviewer.html

 * for UIIntegrationTest_visitors_visitorlog and UIIntegrationTest_ecommerce_log the random failure is always the same and is for the "Goal count block":

![index](https://cloud.githubusercontent.com/assets/466765/8098611/479f6f30-1046-11e5-963c-c84ee57494c3.png)

Maybe we could change the positioning of this element, eg. position it relative to the right column, and this could prevent the random UI failure?

**Comment by mnapoli** [comment]
@mattab that's exactly what I was working on yesterday ;)

Here is the UI diff: http://builds-artifacts.piwik.org/ui-tests.fix-ui-tests-visitor-log/13240.7/screenshot-diffs/diffviewer.html What I did is disable the float and have 2 rows of icons. I can create a pull request later.

**Comment by mnapoli** [comment]
Here is the change for the visitor log: #8090

**Comment by mattab** [comment]
**Note** - in case we notice that we still have as many random build failures, we should revert this pull request: https://github.com/piwik/piwik/pull/8167

**Comment by mattab** [comment]
Moving to 2.14.1 to make an even stronger push for having Green builds, every day, for our UI screenshot tests. The team is all very much looking forward to reach this beautiful goal!

**Comment by mattab** [comment]
* It seems most of the UI random failures were fixed! this is more than great news for us. :sunny: :heart: 
  * The Opt-out ui test is AFAIK the only one that failed (only once) in the last 10 builds or so http://builds-artifacts.piwik.org/ui-tests.master/13783.7/OptOutForm_safari-opted-out 
 * Build running long 
  * The build takes between 38 and 46 minutes. 
  * We are close from the 50min timeout and will soon have this issue
  * out of last 10 builds, 20% of the them timed out and took more than 50minutes.

**Comment by mattab** [comment]
Created separate issue:  Screenshot test CI job is running too long #8222

**Comment by mattab** [comment]
Opt-out with safari agent, seems to be the last randomly failing UI test: http://builds-artifacts.piwik.org/ui-tests.8290/14030.7/

@diosmosis maybe you have some idea what could cause this UI test to randomly fail (it seems to work more than half of the time)

**Comment by mattab** [comment]
Along with "opt-out with safari use case", one of our newly added UI test for the Maps, randomly fail quite rarely. I've seen this failure 2-3 times, last one being: http://builds-artifacts.piwik.org/ui-tests.master/14138.7/ which was around midnight UTC. I guess these two tests fail when the build runs shortly after midnight UTC.

**Comment by mattab** [comment]
We're really not sure how to fix any remaining random UI issue, but we will keep an eye on these issues.

**Comment by mattab** [comment]
We're really not sure how to fix any remaining random UI issue, but we will keep an eye on these issues.

**Comment by mattab** [comment]
It seems that all random issues are finally fixed!!! OMG Happiness is overwhelming me right now ;-) Well done for sticking with us while this was sorted out (took more than one year and dozens of little issues!)

**Comment by mattab** [comment]
Oh no I was mistaken, random failures still occur.... Maybe the only way going forward will be to use electron instead of phantomjs. We will investigate over next 6 months

**Comment by mattab** [comment]
In Piwik 3 we've upgraded Phantomjs to phantomjs 2 since it was required for compatibility with our heavier use of modern JavaScript (AngularJS), but we still have random screenshot fails. Now tracked in https://github.com/piwik/piwik/issues/10536
